# 習題五

## 匯入套件

In [3]:
import pandas as pd
import statsmodels.api as sm

## 讀入資料

### 兩家股票報酬率

In [4]:
df = pd.read_table(r'C:\Users\s0954\Desktop\財金計量方法\Fama_hw-main\B10733033_王廣和_財金計量方法_三因子模型\problem21.txt')
df.head(5)

zhongxin  zyhy        Date
0      3.87  3.51  2014-01-01
1      3.86  3.52  2014-01-02
2      3.80  3.35  2014-01-03
3      3.74  3.23  2014-01-06
4      3.73  3.20  2014-01-07

### 將日期設為index，方便與其他資料整合

In [5]:
df.index = pd.to_datetime(df.Date)
df.head(5)

zhongxin  zyhy        Date
Date                                  
2014-01-01      3.87  3.51  2014-01-01
2014-01-02      3.86  3.52  2014-01-02
2014-01-03      3.80  3.35  2014-01-03
2014-01-06      3.74  3.23  2014-01-06
2014-01-07      3.73  3.20  2014-01-07

### 中新(?)報酬率資料

In [6]:
zhongxin = df.iloc[:,0]
zhongxin.head(5)

Date
2014-01-01    3.87
2014-01-02    3.86
2014-01-03    3.80
2014-01-06    3.74
2014-01-07    3.73
Name: zhongxin, dtype: float64

### 中原海運報酬率資料

In [7]:
zyhy = df.iloc[:,1]
zyhy.head(5)

Date
2014-01-01    3.51
2014-01-02    3.52
2014-01-03    3.35
2014-01-06    3.23
2014-01-07    3.20
Name: zyhy, dtype: float64

### 三因子模型

In [8]:
df2 = pd.read_table(r'C:\Users\s0954\Desktop\財金計量方法\Fama_hw-main\B10733033_王廣和_財金計量方法_三因子模型\ThreeFactors.txt')
df2.head(5)

MarkettypeID TradingDate  RiskPremium1  RiskPremium2      SMB1      SMB2  \
0        P9710   1992/1/30        0.0098        0.0068 -0.000193 -0.003945   
1        P9710   1992/1/31        0.0138        0.0098  0.007597  0.006065   
2        P9710    1992/2/1        0.0058        0.0058  0.011516  0.009544   
3        P9710    1992/2/2        0.0068        0.0068  0.002621  0.002467   
4        P9710    1992/2/3        0.0058        0.0068  0.006923  0.007519   

       HML1      HML2  
0  0.001878  0.001132  
1 -0.003275 -0.002141  
2 -0.006743 -0.006776  
3  0.004255  0.004255  
4 -0.009798 -0.009623

### 將日期設為index，方便與其他資料整合

In [9]:
df2.index = pd.to_datetime(df2.TradingDate)
df2.head(5)

MarkettypeID TradingDate  RiskPremium1  RiskPremium2      SMB1  \
TradingDate                                                                  
1992-01-30         P9710   1992/1/30        0.0098        0.0068 -0.000193   
1992-01-31         P9710   1992/1/31        0.0138        0.0098  0.007597   
1992-02-01         P9710    1992/2/1        0.0058        0.0058  0.011516   
1992-02-02         P9710    1992/2/2        0.0068        0.0068  0.002621   
1992-02-03         P9710    1992/2/3        0.0058        0.0068  0.006923   

                 SMB2      HML1      HML2  
TradingDate                                
1992-01-30  -0.003945  0.001878  0.001132  
1992-01-31   0.006065 -0.003275 -0.002141  
1992-02-01   0.009544 -0.006743 -0.006776  
1992-02-02   0.002467  0.004255  0.004255  
1992-02-03   0.007519 -0.009798 -0.009623

In [10]:
df2.TradingDate = pd.to_datetime(df2.TradingDate)

## 計算

### 三因子公式

## $ R_i-R_f = \alpha + \beta_0 \space (R_{market}-R_{f} )  + \beta_1 \space (R_{small}-R_{big}) + \beta_2 \space (R_{high}-R_{low})$

$個股報酬-無風險利率 = \alpha(超額報酬) + \beta_0 \times 市場風險因子+ \beta_1 \times 公司規模 + \beta_2 \times 淨值因子$

In [11]:
#過濾2014年的資料
filter1 = df2['TradingDate'] >= "2014-01-01"
filter2 = df2['TradingDate'] <= "2014-12-31"
df2_2014 = df2[filter1 & filter2]
df2_2014.head(5)

MarkettypeID TradingDate  RiskPremium1  RiskPremium2      SMB1  \
TradingDate                                                                  
2014-01-02         P9710  2014-01-02     -0.001081      0.000919 -0.017003   
2014-01-03         P9710  2014-01-03     -0.010081     -0.009081 -0.008988   
2014-01-06         P9710  2014-01-06     -0.018081     -0.020081  0.015599   
2014-01-07         P9710  2014-01-07      0.000919      0.001919 -0.008349   
2014-01-08         P9710  2014-01-08     -0.000081      0.001919 -0.005764   

                 SMB2      HML1      HML2  
TradingDate                                
2014-01-02  -0.014747 -0.017604 -0.019279  
2014-01-03  -0.005061 -0.013391 -0.015440  
2014-01-06   0.015898  0.009912  0.009018  
2014-01-07  -0.007034 -0.014506 -0.015003  
2014-01-08  -0.001121 -0.017683 -0.019687

### 中新的三因子表格合併

In [12]:
zhongxin_df = pd.concat([zhongxin,df2_2014['RiskPremium1'],df2_2014['SMB1'],df2_2014['HML1']],axis = 1)
zhongxin_df= zhongxin_df.dropna(axis=0,how='any')
zhongxin_df.head(5)

zhongxin  RiskPremium1      SMB1      HML1
2014-01-02      3.86     -0.001081 -0.017003 -0.017604
2014-01-03      3.80     -0.010081 -0.008988 -0.013391
2014-01-06      3.74     -0.018081  0.015599  0.009912
2014-01-07      3.73      0.000919 -0.008349 -0.014506
2014-01-08      3.74     -0.000081 -0.005764 -0.017683

In [13]:
result1 = sm.OLS(zhongxin_df['zhongxin'],sm.add_constant(zhongxin_df[['RiskPremium1','SMB1','HML1']])).fit()

C:\Users\s0954\Anaconda3\lib\site-packages\numpy\core\fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


In [14]:
print(result1.summary())

                            OLS Regression Results                            
Dep. Variable:               zhongxin   R-squared:                       0.213
Model:                            OLS   Adj. R-squared:                  0.204
Method:                 Least Squares   F-statistic:                     21.80
Date:                Sun, 02 May 2021   Prob (F-statistic):           1.58e-12
Time:                        00:33:53   Log-Likelihood:                -267.62
No. Observations:                 245   AIC:                             543.2
Df Residuals:                     241   BIC:                             557.2
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const            4.6904      0.047     99.706   

### 結果

$R_{zhongxin} = 4.6904 + 8.9308\space Market Risk Premium+37.9980\space Size Factor-16.4894\space BookToMarket Factor $

**There's no data about risk-free

### 中原海運的三因子表格合併

In [15]:
zyhy_df = pd.concat([zyhy,df2_2014['RiskPremium2'],df2_2014['SMB2'],df2_2014['HML2']],axis = 1)
zyhy_df = zyhy_df.dropna(axis=0,how='any')
zyhy_df.head(5)

zyhy  RiskPremium2      SMB2      HML2
2014-01-02  3.52      0.000919 -0.014747 -0.019279
2014-01-03  3.35     -0.009081 -0.005061 -0.015440
2014-01-06  3.23     -0.020081  0.015898  0.009018
2014-01-07  3.20      0.001919 -0.007034 -0.015003
2014-01-08  3.17      0.001919 -0.001121 -0.019687

In [16]:
result2 = sm.OLS(zyhy_df['zyhy'],sm.add_constant(zyhy_df[['RiskPremium2','SMB2','HML2']])).fit()

In [17]:
print(result2.summary())

                            OLS Regression Results                            
Dep. Variable:                   zyhy   R-squared:                       0.136
Model:                            OLS   Adj. R-squared:                  0.125
Method:                 Least Squares   F-statistic:                     12.60
Date:                Sun, 02 May 2021   Prob (F-statistic):           1.11e-07
Time:                        00:33:54   Log-Likelihood:                -367.42
No. Observations:                 245   AIC:                             742.8
Df Residuals:                     241   BIC:                             756.8
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const            3.7480      0.071     53.055   

### 結果

$R_{zyhy} = 3.7480 + 15.2256\space Market Risk Premium+39.7967\space Size Factor$

**BookToMarket Factor is not statistically significant

**There's no data about risk-free

# 習題六

In [18]:
df3 = pd.read_csv(r'C:\Users\s0954\Desktop\財金計量方法\Fama_hw-main\B10733033_王廣和_財金計量方法_三因子模型\2498.csv')
df3.head(5)

Date      Date.1   Close    Open    High     Low  Volume     ROI
0  2014-01-02  2014-01-02  137.58  140.56  141.06  137.09    7964 -1.7730
1  2014-01-03  2014-01-03  137.58  137.58  140.07  137.09    6599  0.0000
2  2014-01-06  2014-01-06  132.62  137.58  138.58  132.12   14005 -3.6101
3  2014-01-07  2014-01-07  130.13  131.62  132.62  128.64   13705 -1.8727
4  2014-01-08  2014-01-08  129.14  130.13  131.62  129.14    7623 -0.7634

註:這邊直接取用ROI 

# $ROI = \frac{Close_{t}-Close_{t-1}}{Close_{t-1}}$

In [19]:
df3.index = pd.to_datetime(df3.Date)
df3.Date = pd.to_datetime(df3.Date)
df3.head(5)

Date      Date.1   Close    Open    High     Low  Volume  \
Date                                                                        
2014-01-02 2014-01-02  2014-01-02  137.58  140.56  141.06  137.09    7964   
2014-01-03 2014-01-03  2014-01-03  137.58  137.58  140.07  137.09    6599   
2014-01-06 2014-01-06  2014-01-06  132.62  137.58  138.58  132.12   14005   
2014-01-07 2014-01-07  2014-01-07  130.13  131.62  132.62  128.64   13705   
2014-01-08 2014-01-08  2014-01-08  129.14  130.13  131.62  129.14    7623   

               ROI  
Date                
2014-01-02 -1.7730  
2014-01-03  0.0000  
2014-01-06 -3.6101  
2014-01-07 -1.8727  
2014-01-08 -0.7634

In [20]:
df_2498 = pd.concat([df3['ROI'],
                     (df2_2014['RiskPremium1']+df2_2014['RiskPremium2'])/2,
                     (df2_2014['SMB1']+df2_2014['SMB2'])/2,
                     (df2_2014['HML1']+df2_2014['HML2'])/2
                    ],axis = 1)
df_2498 = df_2498.dropna(axis=0,how='any')
df_2498.columns= ['ROI','RiskPremium','SMB','HML']
df_2498.head(5)

ROI  RiskPremium       SMB       HML
2014-01-02 -1.7730    -0.000081 -0.015875 -0.018441
2014-01-03  0.0000    -0.009581 -0.007024 -0.014416
2014-01-06 -3.6101    -0.019081  0.015748  0.009465
2014-01-07 -1.8727     0.001419 -0.007691 -0.014755
2014-01-08 -0.7634     0.000919 -0.003443 -0.018685

In [26]:
#2498三因子模型
result3 = sm.OLS(df_2498['ROI'],sm.add_constant(df_2498[['RiskPremium','SMB','HML']])).fit()
print(result3.summary())

                            OLS Regression Results                            
Dep. Variable:                    ROI   R-squared:                       0.019
Model:                            OLS   Adj. R-squared:                  0.007
Method:                 Least Squares   F-statistic:                     1.519
Date:                Sun, 02 May 2021   Prob (F-statistic):              0.210
Time:                        00:36:01   Log-Likelihood:                -506.41
No. Observations:                 238   AIC:                             1021.
Df Residuals:                     234   BIC:                             1035.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const          -0.0136      0.134     -0.101      

C:\Users\s0954\Anaconda3\lib\site-packages\numpy\core\fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


In [28]:
result3.params

const          -0.013582
RiskPremium    26.005860
SMB            -9.990303
HML             6.632819
dtype: float64

In [32]:
#過濾2015年的資料
filter3 = df2['TradingDate'] >= "2015-01-01"
filter4 = df2['TradingDate'] <= "2015-01-31"
df_2015 = df2[filter3 & filter4]
df_2015.head(5)

MarkettypeID TradingDate  RiskPremium1  RiskPremium2      SMB1  \
TradingDate                                                                  
2015-01-05         P9710  2015-01-05      0.033926      0.030926  0.037147   
2015-01-06         P9710  2015-01-06      0.002926      0.004926 -0.021998   
2015-01-07         P9710  2015-01-07      0.002926      0.002926  0.007548   
2015-01-08         P9710  2015-01-08     -0.022074     -0.019074 -0.025485   
2015-01-09         P9710  2015-01-09     -0.002074     -0.002074  0.007416   

                 SMB2      HML1      HML2  
TradingDate                                
2015-01-05   0.032969  0.033216  0.036105  
2015-01-06  -0.018519 -0.023558 -0.025566  
2015-01-07   0.006314  0.005183  0.005311  
2015-01-08  -0.022480 -0.029721 -0.031789  
2015-01-09   0.007961  0.002365  0.001750

In [40]:
RP_2015 = (df_2015['RiskPremium1']+df_2015['RiskPremium2'])/2
SMB_2015 = (df_2015['SMB1']+df_2015['SMB2'])/2
HML_2015 = (df_2015['HML1']+df_2015['HML2'])/2

In [42]:
a,b1,b2,b3 = result3.params

In [43]:
y = a+b1*RP_2015+b2*SMB_2015+b3*HML_2015
y

TradingDate
2015-01-05    0.709341
2015-01-06    0.127991
2015-01-07    0.028071
2015-01-08   -0.513026
2015-01-09   -0.130681
2015-01-12   -0.614656
2015-01-13    0.167182
2015-01-14   -0.056077
2015-01-15    0.698264
2015-01-16    0.246169
2015-01-19   -1.706403
2015-01-20    0.573877
2015-01-21    1.054758
2015-01-22    0.172720
2015-01-23   -0.046286
2015-01-26    0.434991
2015-01-27   -0.188071
2015-01-28   -0.234587
2015-01-29   -0.282689
2015-01-30   -0.328223
dtype: float64

### 結果

$R_{2498} = 26.0059\space MarketRiskPremium$

**There's no data about risk-free

**Neither Size Factor nor BookToMarket Factor is not statistically significant

In [22]:
#2498CAPM模型
result3_1 = sm.OLS(df_2498['ROI'],sm.add_constant(df_2498[['RiskPremium']])).fit()
print(result3_1.summary())

                            OLS Regression Results                            
Dep. Variable:                    ROI   R-squared:                       0.018
Model:                            OLS   Adj. R-squared:                  0.014
Method:                 Least Squares   F-statistic:                     4.329
Date:                Sun, 02 May 2021   Prob (F-statistic):             0.0385
Time:                        00:33:55   Log-Likelihood:                -506.54
No. Observations:                 238   AIC:                             1017.
Df Residuals:                     236   BIC:                             1024.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const          -0.0150      0.134     -0.112      

### 結果

$R_{2498} = 25.0934\space MarketRiskPremium$

**There's no data about risk-free

In [45]:
c,d1 = result3_1.params

In [46]:
y2 = c +d1*RP_2015
y2

TradingDate
2015-01-05    0.798632
2015-01-06    0.083470
2015-01-07    0.058376
2015-01-08   -0.531320
2015-01-09   -0.067091
2015-01-12   -0.380759
2015-01-13    0.096016
2015-01-14   -0.092185
2015-01-15    0.748446
2015-01-16    0.296764
2015-01-19   -1.785991
2015-01-20    0.522605
2015-01-21    1.099754
2015-01-22    0.146203
2015-01-23   -0.029451
2015-01-26    0.560245
2015-01-27   -0.205105
2015-01-28   -0.330572
2015-01-29   -0.305479
2015-01-30   -0.368212
dtype: float64

In [50]:
predict = pd.concat([y,y2],axis=1)


<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 20 entries, 2015-01-05 to 2015-01-30
Data columns (total 2 columns):
0    20 non-null float64
1    20 non-null float64
dtypes: float64(2)
memory usage: 480.0 bytes


In [51]:
from scipy import stats

In [52]:
stats.describe(predict)

DescribeResult(nobs=20, minmax=(array([-1.70640307, -1.78599139]), array([1.05475799, 1.09975366])), mean=array([0.00563324, 0.01571726]), variance=array([0.34789154, 0.37524649]), skewness=array([-0.88388027, -0.88195965]), kurtosis=array([1.88249214, 2.07807346]))